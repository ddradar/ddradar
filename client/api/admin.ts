import type { NuxtHTTPInstance } from '@nuxt/http'

import type { SongInfo } from '~/api/song'
import type { NotificationSchema } from '~/core/db/notification'

type NotificationRequest = Omit<
  NotificationSchema,
  'id' | 'sender' | 'timeStamp'
> & {
  id?: string
  timeStamp?: number
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postNotification
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: NotificationRequest
) {
  return $http.$post<NotificationRequest>('/api/v1/admin/notification', {
    ...body,
    sender: 'SYSTEM',
  })
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postSongInfo
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: SongInfo
) {
  return $http.$post<SongInfo>('/api/v1/admin/songs', body)
}
