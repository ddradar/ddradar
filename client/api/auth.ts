import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

type AuthResult = {
  clientPrincipal: Api.ClientPrincipal | null
}

/**
 * Call "Get Credentials" API.
 * @see https://docs.microsoft.com/azure/static-web-apps/user-information
 */
export async function getClientPrincipal(
  $http: Pick<NuxtHTTPInstance, '$get'>
) {
  const { clientPrincipal } = await $http.$get<AuthResult>('/.auth/me')
  return clientPrincipal
}
