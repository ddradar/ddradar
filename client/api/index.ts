export const apiPrefix = '/api/v1'
