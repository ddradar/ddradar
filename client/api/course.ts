import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function getCourseType(type: number) {
  return type === 1 ? 'NONSTOP' : type === 2 ? '段位認定' : ''
}

/**
 * Call "Get Course List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses/
 */
export function getCourseList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series?: 16 | 17,
  type?: 1 | 2
) {
  const searchParams = new URLSearchParams()
  if (series) searchParams.append('series', `${series}`)
  if (type) searchParams.append('type', `${type}`)
  return $http.$get<Api.CourseListData[]>(`${apiPrefix}/courses`, {
    searchParams,
  })
}

/**
 * Call "Get Course Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses__id/
 */
export function getCourseInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.CourseInfo>(`${apiPrefix}/courses/${id}`)
}
