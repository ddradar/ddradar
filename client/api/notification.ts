import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Notification List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification/
 */
export function getNotificationList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  topOnly?: boolean
) {
  const query = topOnly ? '?scope=top' : ''
  return $http.$get<Api.NotificationListData[]>(
    `${apiPrefix}/notification${query}`
  )
}

/**
 * Call "Get Notification Info" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification__id/
 */
export function getNotificationInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.NotificationInfo>(`${apiPrefix}/notification/${id}`)
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification--post/
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: Omit<Api.NotificationBody, 'sender'>
) {
  return $http.$post<Api.NotificationBody>(`${apiPrefix}/notification`, {
    ...body,
    sender: 'SYSTEM',
  })
}
