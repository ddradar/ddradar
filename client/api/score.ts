import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--get/
 */
export function getChartScore(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  scope?: 'private' | 'medium' | 'full'
) {
  const query = scope ? `?scope=${scope}` : ''
  return $http.$get<Api.ScoreInfo[]>(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}${query}`
  )
}

/**
 * Call "Post Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--post/
 */
export function postChartScore(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  score: Api.ScoreBody
) {
  return $http.$post(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}`,
    score
  )
}

/**
 * Call "Delete Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--delete/
 */
export function deleteChartScore(
  $http: Pick<NuxtHTTPInstance, 'delete'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4
) {
  return $http.delete(`/api/v1/scores/${songId}/${playStyle}/${difficulty}`)
}

/**
 * Call "Post Song Scores" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id--post/
 */
export function postSongScores(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  scores: Api.ScoreListBody[]
) {
  return $http.$post<Api.ScoreInfo[]>(`${apiPrefix}/scores/${songId}`, scores)
}
