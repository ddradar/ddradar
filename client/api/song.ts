import type { Api } from '@ddradar/core'
import { Song } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function shortenSeriesName(series: string) {
  return series.replace(/^(DDR |DanceDanceRevolution )\(?([^)]+)\)?$/, '$2')
}

export function getChartTitle({
  playStyle,
  difficulty,
  level,
}: Pick<Api.ChartInfo, 'playStyle' | 'difficulty' | 'level'>) {
  const shortPlayStyle = `${Song.playStyleMap.get(playStyle)![0]}P` as
    | 'SP'
    | 'DP'
  const difficultyName = Song.difficultyMap.get(difficulty)!
  return `${shortPlayStyle}-${difficultyName} (${level})`
}

export function getDisplayedBPM({
  minBPM,
  maxBPM,
}: Pick<Api.SongInfo, 'minBPM' | 'maxBPM'>) {
  if (!minBPM || !maxBPM) return '???'
  if (minBPM === maxBPM) return `${minBPM}`
  return `${minBPM}-${maxBPM}`
}

/**
 * Call "Get Song Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__id/
 */
export function getSongInfo($http: Pick<NuxtHTTPInstance, '$get'>, id: string) {
  return $http.$get<Api.SongInfo>(`${apiPrefix}/songs/${id}`)
}

/**
 * Call "Search Song by Name" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__name__id/
 */
export function searchSongByName(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  nameIndex: number
) {
  return $http.$get<Api.SongListData[]>(`${apiPrefix}/songs/name/${nameIndex}`)
}

/**
 * Call "Search Song by Series" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs__series__id/
 */
export function searchSongBySeries(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series: number
) {
  return $http.$get<Api.SongListData[]>(`${apiPrefix}/songs/series/${series}`)
}

/**
 * Call "Search Charts" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/charts__style__level/
 */
export function searchCharts(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  playStyle: Song.PlayStyle,
  level: number
) {
  return $http.$get<Api.ChartInfo[]>(
    `${apiPrefix}/charts/${playStyle}/${level}`
  )
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/songs--post/
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: Api.SongInfo
) {
  return $http.$post<Api.SongInfo>(`${apiPrefix}/songs`, body)
}
