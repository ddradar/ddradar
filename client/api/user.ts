import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "User Exists" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/user__exists__id/
 */
export async function existsUser(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  const { exists } = await $http.$get<Api.ExistsUser>(
    `${apiPrefix}/user/exists/${id}`
  )
  return exists
}

/**
 * Call "Get Current User Data" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/user--get/
 */
export function getCurrentUser($http: Pick<NuxtHTTPInstance, '$get'>) {
  return $http.$get<Api.CurrentUserInfo>(`${apiPrefix}/user`)
}

/**
 * Call "Get User List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/users/
 */
export function getUserList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  name?: string,
  area?: number,
  code?: number
) {
  const searchParams = new URLSearchParams()
  if (name) searchParams.append('name', name)
  if (area) searchParams.append('area', `${area}`)
  if (code) searchParams.append('code', `${code}`)
  return $http.$get<Api.UserInfo[]>(`${apiPrefix}/users`, { searchParams })
}

/**
 * Call "Get User Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/users__id/
 */
export function getUserInfo($http: Pick<NuxtHTTPInstance, '$get'>, id: string) {
  return $http.$get<Api.UserInfo>(`${apiPrefix}/users/${id}`)
}

/**
 * Call "Get Clear Status" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/users__id__clear/
 */
export function getClearStatus(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  userId: string,
  playStyle: 1 | 2
) {
  return $http.$get<Api.ClearStatus[]>(
    `${apiPrefix}/users/${userId}/clear?playStyle=${playStyle}`
  )
}

/**
 * Call "Get Score Status" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/users__id__score/
 */
export function getScoreStatus(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  userId: string,
  playStyle: 1 | 2
) {
  return $http.$get<Api.ScoreStatus[]>(
    `${apiPrefix}/users/${userId}/score?playStyle=${playStyle}`
  )
}

/**
 * Call "Get Groove Radar" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/users__id__radar__style/
 */
export function getGrooveRadar(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  userId: string,
  playStyle: 1 | 2
) {
  return $http.$get<Api.GrooveRadarInfo[]>(
    `${apiPrefix}/users/${userId}/radar/${playStyle}`
  )
}

/**
 * Call "Post User Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/user--post/
 */
export function postUserInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  user: Api.CurrentUserInfo
) {
  return $http.$post<Api.CurrentUserInfo>(`${apiPrefix}/user`, user)
}
