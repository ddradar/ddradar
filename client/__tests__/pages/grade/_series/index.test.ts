import type { Context } from '@nuxt/types'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'
import VueI18n from 'vue-i18n'

import { getCourseList } from '~/api/course'
import GradeListPage from '~/pages/grade/_series/index.vue'

jest.mock('~/api/course')
const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(VueI18n)

describe('pages/grade/_series/index.vue', () => {
  const $fetchState = { pending: false }
  describe.each(['ja', 'en'])('{ locale: %s } snapshot test', locale => {
    const i18n = new VueI18n({ locale, silentFallbackWarn: true })
    test('renders correctly', () => {
      // Arrange
      const $route = { params: { series: '16' }, path: '/grade/16' }
      const mocks = { $fetchState, $route }
      const wrapper = shallowMount(GradeListPage, { localVue, mocks, i18n })

      // Act - Assert
      expect(wrapper).toMatchSnapshot()
    })
  })

  // Lifecycle
  describe('validate()', () => {
    const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
    test.each(['', 'foo', '0', '1', '10'])('/%s returns false', series => {
      // Arrange
      const mocks = { $fetchState, $route: { params: { series } } }
      const wrapper = shallowMount(GradeListPage, { localVue, mocks, i18n })
      const ctx = { ...mocks.$route } as unknown as Context

      // Act - Assert
      expect(wrapper.vm.$options.validate!(ctx)).toBe(false)
    })
    test.each(['16', '17'])('/%s returns true', series => {
      // Arrange
      const mocks = { $fetchState, $route: { params: { series } } }
      const wrapper = shallowMount(GradeListPage, { localVue, mocks, i18n })
      const ctx = { ...mocks.$route } as unknown as Context

      // Act - Assert
      expect(wrapper.vm.$options.validate!(ctx)).toBe(true)
    })
  })
  describe('fetch()', () => {
    const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
    const apiMock = mocked(getCourseList)
    beforeEach(() => {
      apiMock.mockClear()
      apiMock.mockResolvedValue([])
    })
    test.each([
      ['16', 16],
      ['17', 17],
    ])('/%s calls getCourseList($http, %i, 2)', async (series, expected) => {
      // Arrange
      const $route = { params: { series }, path: `/grade/${series}` }
      const mocks = { $fetchState, $http: { $get: jest.fn() }, $route }
      const wrapper = shallowMount(GradeListPage, { localVue, mocks, i18n })

      // Act
      // @ts-ignore
      await wrapper.vm.$options.fetch?.call(wrapper.vm)

      // Assert
      expect(apiMock).toBeCalledTimes(1)
      expect(apiMock).toBeCalledWith(mocks.$http, expected, 2)
    })
  })

  // Computed
  describe('get title()', () => {
    test.each([
      ['16', 'ja', 'DanceDanceRevolution A20 - 段位認定'],
      ['17', 'ja', 'DanceDanceRevolution A20 PLUS - 段位認定'],
      ['16', 'en', 'DanceDanceRevolution A20 - GRADE'],
      ['17', 'en', 'DanceDanceRevolution A20 PLUS - GRADE'],
    ])('/%s (locale:%s) returns "%s"', (series, locale, expected) => {
      const i18n = new VueI18n({ locale, silentFallbackWarn: true })
      // Arrange
      const $route = { params: { series }, path: `/grade/${series}` }
      const mocks = { $fetchState, $route }
      const wrapper = shallowMount(GradeListPage, { localVue, mocks, i18n })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.title).toBe(expected)
    })
  })
})
