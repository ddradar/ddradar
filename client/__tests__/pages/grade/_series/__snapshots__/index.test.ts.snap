// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/grade/_series/index.vue { locale: "en" } snapshot test renders correctly 1`] = `
<section class="section">
  <h1 class="title">DanceDanceRevolution A20 - GRADE</h1>
  <div class="buttons"><a class="button is-info">
      <!----><span>
      NONSTOP(A20)
    </span>
      <!---->
    </a><button type="button" to="/grade/16" disabled="disabled" class="button is-info is-outlined">
      <!----><span>
      GRADE(A20)
    </span>
      <!---->
    </button><a class="button is-info">
      <!----><span>
      NONSTOP(A20 PLUS)
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      GRADE(A20 PLUS)
    </span>
      <!---->
    </a></div>
  <courselist-stub courses=""></courselist-stub>
</section>
`;

exports[`pages/grade/_series/index.vue { locale: "ja" } snapshot test renders correctly 1`] = `
<section class="section">
  <h1 class="title">DanceDanceRevolution A20 - 段位認定</h1>
  <div class="buttons"><a class="button is-info">
      <!----><span>
      NONSTOP(A20)
    </span>
      <!---->
    </a><button type="button" to="/grade/16" disabled="disabled" class="button is-info is-outlined">
      <!----><span>
      段位認定(A20)
    </span>
      <!---->
    </button><a class="button is-info">
      <!----><span>
      NONSTOP(A20 PLUS)
    </span>
      <!---->
    </a><a class="button is-info">
      <!----><span>
      段位認定(A20 PLUS)
    </span>
      <!---->
    </a></div>
  <courselist-stub courses=""></courselist-stub>
</section>
`;
